srml/balances/src/tests.rs

```diff
@@ -21,7 +21,110 @@
 use super::*;
 use mock::{Balances, ExtBuilder, Runtime, System};
 use runtime_io::with_externalities;
-use srml_support::{assert_noop, assert_ok, assert_err};
+use srml_support::{
+	assert_noop, assert_ok, assert_err,
+	traits::{LockableCurrency, LockIdentifier, WithdrawReason, WithdrawReasons, TransferAsset}
+};
+
+const ID_1: LockIdentifier = *b"1       ";
+const ID_2: LockIdentifier = *b"2       ";
+const ID_3: LockIdentifier = *b"3       ";
+
+#[test]
+fn basic_locking_should_work() {
+	with_externalities(
+		&mut ExtBuilder::default()
+			.existential_deposit(10)
+			.creation_fee(50)
+			.monied(true)
+			.build(),
+		|| {
+			assert_eq!(Balances::free_balance(&1), 2560);
+			Balances::set_lock(ID_1, &1, 2500, u64::max_value(), WithdrawReasons::all());
+			assert_noop!(<Balances as TransferAsset<_>>::transfer(&1, &2, 100), "account liquidity restrictions prevent withdrawal");
+		}
+	);
+}
+
+#[test]
+fn partial_locking_should_work() {
+	with_externalities(
+		&mut ExtBuilder::default()
+			.existential_deposit(10)
+			.creation_fee(50)
+			.monied(true)
+			.build(),
+		|| {
+			Balances::set_lock(ID_1, &1, 1280, u64::max_value(), WithdrawReasons::all());
+			assert_ok!(<Balances as TransferAsset<_>>::transfer(&1, &2, 100));
+		}
+	);
+}
+
+#[test]
+fn lock_removal_should_work() {
+	with_externalities(
+		&mut ExtBuilder::default()
+			.existential_deposit(10)
+			.creation_fee(50)
+			.monied(true)
+			.build(),
+		|| {
+			Balances::set_lock(ID_1, &1, u64::max_value(), u64::max_value(), WithdrawReasons::all());
+			Balances::remove_lock(ID_1, &1);
+			assert_ok!(<Balances as TransferAsset<_>>::transfer(&1, &2, 100));
+		}
+	);
+}
+
+#[test]
+fn lock_replacement_should_work() {
+	with_externalities(
+		&mut ExtBuilder::default()
+			.existential_deposit(10)
+			.creation_fee(50)
+			.monied(true)
+			.build(),
+		|| {
+			Balances::set_lock(ID_1, &1, u64::max_value(), u64::max_value(), WithdrawReasons::all());
+			Balances::set_lock(ID_1, &1, 1280, u64::max_value(), WithdrawReasons::all());
+			assert_ok!(<Balances as TransferAsset<_>>::transfer(&1, &2, 100));
+		}
+	);
+}
+
+#[test]
+fn double_locking_should_work() {
+	with_externalities(
+		&mut ExtBuilder::default()
+			.existential_deposit(10)
+			.creation_fee(50)
+			.monied(true)
+			.build(),
+		|| {
+			Balances::set_lock(ID_1, &1, 1280, u64::max_value(), WithdrawReasons::all());
+			Balances::set_lock(ID_2, &1, 1280, u64::max_value(), WithdrawReasons::all());
+			assert_ok!(<Balances as TransferAsset<_>>::transfer(&1, &2, 100));
```

Comment:
So I think I am missing something about the nature of the lock here that this strikes me as counter-intuitive. To me, this account should now have all of its 2*1280 units of fund locked and the transfer should fail.

Reply:
following up on the above: shouldn't a lock affect free-vesting balance? 
In other words what would have made this test (of course for me) more understandable was to see the operation and then a query afterwards on how it affected the different balance storage items corresponding to `1`: 

```
Balances::set_lock(ID_1, &1, 1280, u64::max_value(), WithdrawReasons::all());
Balances::set_lock(ID_2, &1, 1280, u64::max_value(), WithdrawReasons::all());

// knowing that initially free = 2580 and vesting = 0
assert_eq!(Balances::free_balance(&1), XXX); 
assert_eq!(Balances::vesting_balance(&1), YYY); 
```

Reply:
@joepetrowski since I see that you are assigned to the documentation of this module this might also be related to you.